dpnp/fft/dpnp_utils_fft.py

```diff
@@ -0,0 +1,318 @@
+The main computational work is performed by enabling FFT functions
+available as a pybind11 extension.
+
+"""
+
+# pylint: disable=protected-access
+# pylint: disable=c-extension-no-member
+# pylint: disable=no-name-in-module
+
+import dpctl
+import dpctl.tensor as dpt
+import dpctl.tensor._tensor_impl as ti
+import dpctl.utils as dpu
+import numpy
+from dpctl.utils import ExecutionPlacementError
+from numpy.core.numeric import normalize_axis_index
+
+import dpnp
+import dpnp.backend.extensions.fft._fft_impl as fi
+
+from ..dpnp_array import dpnp_array
+from ..dpnp_utils import map_dtype_to_device
+from ..dpnp_utils.dpnp_utils_linearalgebra import (
+    _standardize_strides_to_nonzero,
+)
+
+__all__ = [
+    "dpnp_fft",
+]
+
+
+def _check_norm(norm):
+    if norm not in (None, "ortho", "forward", "backward"):
+        raise ValueError(
+            f"Invalid norm value {norm} should be None, "
+            '"ortho", "forward", or "backward".'
+        )
+
+
+def _commit_descriptor(a, in_place, a_strides, index, axes):
+    """Commit the FFT descriptor for the input array."""
+
+    a_shape = a.shape
+    shape = a_shape[index:]
+    strides = (0,) + a_strides[index:]
+    if a.dtype == dpnp.complex64:
+        dsc = fi.Complex64Descriptor(shape)
+    else:
+        dsc = fi.Complex128Descriptor(shape)
+
+    dsc.fwd_strides = strides
+    dsc.bwd_strides = dsc.fwd_strides
+    dsc.transform_in_place = in_place
+    if axes is not None:  # batch_fft
+        dsc.fwd_distance = a_strides[0]
+        dsc.bwd_distance = dsc.fwd_distance
+        dsc.number_of_transforms = numpy.prod(a_shape[0])
+    dsc.commit(a.sycl_queue)
+
+    return dsc
+
+
+def _compute_result(dsc, a, out, forward, a_strides):
+    """Compute the result of the FFT."""
+
+    exec_q = a.sycl_queue
+    _manager = dpu.SequentialOrderManager[exec_q]
+    dep_evs = _manager.submitted_events
+
+    a_usm = dpnp.get_usm_ndarray(a)
+    if dsc.transform_in_place:
+        # in-place transform
+        ht_fft_event, fft_event = fi._fft_in_place(
+            dsc, a_usm, forward, depends=dep_evs
+        )
+        res_usm = a_usm
+    else:
+        if (
+            out is not None
+            and out.strides == a_strides
+            and not ti._array_overlap(a_usm, dpnp.get_usm_ndarray(out))
+        ):
+            res_usm = dpnp.get_usm_ndarray(out)
+        else:
+            # Result array that is used in OneMKL must have the exact same
+            # stride as input array
+            res_usm = dpt.usm_ndarray(
+                a.shape,
+                dtype=a.dtype,
+                buffer=a.usm_type,
+                strides=a_strides,
+                offset=0,
+                buffer_ctor_kwargs={"queue": a.sycl_queue},
+            )
+        ht_fft_event, fft_event = fi._fft_out_of_place(
+            dsc, a_usm, res_usm, forward, depends=dep_evs
+        )
+    _manager.add_event_pair(ht_fft_event, fft_event)
+
+    return res_usm
```

Comment:
I believe it would be clear to always return dpnp array, otherwise there will be a mix between dpnp array and usm_ndarray for different functions and it might confuse and lead to errors.
Or we need somehow explicitly point out that usm_ndarray will be return by the function name or similar.